zproject/backends.py

```diff
@@ -910,6 +943,78 @@ def sync_custom_profile_fields_from_ldap(
         except SyncUserError as e:
             raise ZulipLDAPError(str(e)) from e
 
+    def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser) -> None:
+        """
+        For the groups set up for syncing for the realm in LDAP_GROUP_SYNC_LIST:
+
+        (1) Makes sure the user has membership in the Zulip UserGroups corresponding
+            to the LDAP groups ldap_user belongs to.
+        (2) Makes sure the user doesn't have membership in the Zulip UserGroups corresponding
+            to the LDAP groups ldap_user doesn't belong to.
+        """
+
+        if user_profile.realm.string_id not in settings.LDAP_GROUP_SYNC_LIST:
+            # no groups to sync for this realm
+            return
+
```

Comment:
@mateuszmandera should this function have `transaction.atomic` on it? I think the answer is no -- it's fine if one of the two transactions (to bulk-add or bulk-remove) fails -- running it again will recover cleanly -- and both of those individual functions run in transactions.

Reply:
The functions are intentionally only doing "changes", no full resets of user groups. So if something fails, there should not be any unwanted state.